### ISSUE TITLE
Implement full GoogLeNet backward pass (~3500 lines)

### ISSUE THREAD

**Issue by mvillmow** [HomericIntelligence/ProjectOdyssey#3295]
The backward pass in examples/googlenet-cifar10/train.mojo is a documented placeholder (issue #3184). Requires: 9 Inception module backward passes (4-way gradient split from concatenation, 4 branch backward paths each), auxiliary classifier backward pass, and SGD+momentum parameter updates. Each Inception module backward requires splitting gradients from the depth-concatenation operation across all 4 branches.

_Follow-up from #3084_

**Comment by mvillmow** [comment]
# Implementation Plan

Now I have enough context to write a comprehensive implementation plan.

---

# Implementation Plan: GoogLeNet Full Backward Pass (Issue #3295)

## 1. Objective

Replace the placeholder backward pass in `examples/googlenet-cifar10/train.mojo` with a complete, working implementation (~3500 lines). This includes:

- A cached forward pass that saves all intermediate activations for each of the 9 Inception modules
- Full backward pass through all 9 Inception modules (4-branch gradient split per module)
- Backward pass through the initial conv block, MaxPool layers, GAP, dropout, and FC layer
- SGD+momentum parameter update for all ~200 parameters
- Layerwise backward tests for the Inception module backward pass

---

## 2. Approach

### Key Design Decisions

1. **Split `train_epoch` into a standalone `compute_gradients` function** — matching the AlexNet pattern (`examples/alexnet-cifar10/train.mojo:56`). This avoids the closure-captures-non-copyable-struct Mojo limitation.

2. **Explicit activation cache struct** — Create `GoogLeNetCache` struct to hold all intermediate tensors needed for the backward pass (~100 cached tensors across 9 Inception modules + outer layers). Struct fields are `ExTensor` (not `List[ExTensor]`) for safe ownership.

3. **Separate `inception_module_backward` helper** — A standalone function `fn inception_module_backward(...)` taking the module's weights, cached activations, and incoming gradient, returning accumulated input gradient + all parameter gradients. This avoids 9 copies of ~300 lines of branch backward code.

4. **Depth-concat gradient split** — Mirror `concatenate_depthwise` in `model.mojo:522-588`. The inverse is a channel-slice operation using `ExTensor.slice()` on axis=1 at offsets `[0, c1, c1+c2, c1+c2+c3]`.

5. **SGD+momentum pattern** — Match `examples/alexnet-cifar10/train.mojo:265-337`: extract velocities from `List[ExTensor]` by index, compute `vel = momentum * vel + lr * grad`, update `weight -= vel`, write back.

6. **Batch norm backward** — Use `batch_norm2d_backward(grad, x, gamma, running_mean, running_var, training=True)` returning `Tuple[ExTensor, ExTensor, ExTensor]` (grad_input, grad_gamma, grad_beta) — per `shared/core/normalization.mojo:463-471`.

7. **Conv2d backward** — Use `conv2d_backward(grad, input, weights, stride, padding)` returning `GradientTriple` with fields `.grad_input`, `.grad_weights`, `.grad_bias` — per `shared/core/gradient_types.mojo:68-79`.

---

## 3. Files to Create

None new. All changes are modifications to existing files.

---

## 4. Files to Modify

### `examples/googlenet-cifar10/train.mojo` — Primary implementation file

**Remove:**
- All comment blocks documenting the placeholder (lines 97-253 in the current file)
- The `NOTE:` print statements in `main()` (lines 436-441)
- The `compute_batch_loss` closure inside `train_epoch` (not compatible with capturing `mut model`)

**Add:**

#### 4a. `struct InceptionModuleCache` (before `train_epoch`)

```mojo
struct InceptionModuleCache:
    """Cached activations for one Inception module's backward pass."""
    # Branch 1 intermediates
    var b1_conv_out: ExTensor    # after conv1x1_1
    var b1_bn_out: ExTensor      # after bn1x1_1
    var b1_relu_out: ExTensor    # after relu (= final b1)
    # Branch 2 intermediates
    var b2_conv1_out: ExTensor   # after conv1x1_2
    var b2_bn1_out: ExTensor     # after bn1x1_2
    var b2_relu1_out: ExTensor   # after first relu
    var b2_conv2_out: ExTensor   # after conv3x3
    var b2_bn2_out: ExTensor     # after bn3x3
    var b2_relu2_out: ExTensor   # after second relu (= final b2)
    # Branch 3 intermediates
    var b3_conv1_out: ExTensor   # after conv1x1_3
    var b3_bn1_out: ExTensor     # after bn1x1_3
    var b3_relu1_out: ExTensor   # after first relu
    var b3_conv2_out: ExTensor   # after conv5x5
    var b3_bn2_out: ExTensor     # after bn5x5
    var b3_relu2_out: ExTensor   # after second relu (= final b3)
    # Branch 4 intermediates
    var b4_pool_out: ExTensor    # after maxpool
    var b4_conv_out: ExTensor    # after conv1x1_4
    var b4_bn_out: ExTensor      # after bn1x1_4
    var b4_relu_out: ExTensor    # after relu (= final b4)
    # Module input (needed for all branch backward conv inputs)
    var module_input: ExTensor
```

#### 4b. `struct GoogLeNetCache` (before `train_epoch`)

```mojo
struct GoogLeNetCache:
    """All cached activations for GoogLeNet forward pass."""
    # Initial block
    var initial_conv_out: ExTensor
    var initial_bn_out: ExTensor
    var initial_relu_out: ExTensor
    var pool1_out: ExTensor         # after first maxpool → input to 3a
    # 9 inception module caches
    var cache_3a: InceptionModuleCache
    var cache_3b: InceptionModuleCache
    var pool2_out: ExTensor         # after second maxpool → input to 4a
    var cache_4a: InceptionModuleCache
    var cache_4b: InceptionModuleCache
    var cache_4c: InceptionModuleCache
    var cache_4d: InceptionModuleCache
    var cache_4e: InceptionModuleCache
    var pool3_out: ExTensor         # after third maxpool → input to 5a
    var cache_5a: InceptionModuleCache
    var cache_5b: InceptionModuleCache
    # Classifier
    var gap_out: ExTensor
    var flattened_out: ExTensor
    var dropout_mask: ExTensor
    var logits: ExTensor
```

#### 4c. `fn forward_with_cache` (replaces `model.forward()` in training path)

A new function (not modifying the model struct) that runs the forward pass identically to `GoogLeNet.forward()` but returns both logits and the full cache:

```mojo
fn forward_with_cache(
    mut model: GoogLeNet,
    x: ExTensor,
) raises -> Tuple[ExTensor, GoogLeNetCache]:
```

Each Inception module's forward pass is re-run inline (not via `model.inception_Xa.forward()`) to capture intermediate tensors before they're discarded.

#### 4d. `fn inception_module_backward` helper

```mojo
fn inception_module_backward(
    inception: InceptionModule,
    cache: InceptionModuleCache,
    grad_concat_out: ExTensor,    # gradient of concatenated output
    out_1x1: Int,
    out_3x3: Int,
    out_5x5: Int,
    pool_proj: Int,
) raises -> Tuple[
    ExTensor,   # grad_x (input gradient for previous layer)
    # Branch 1 param grads: conv weights, bias, bn gamma, bn beta
    ExTensor, ExTensor, ExTensor, ExTensor,
    # Branch 2 param grads: reduce conv w/b, bn1 gamma/beta, 3x3 w/b, bn2 gamma/beta
    ExTensor, ExTensor, ExTensor, ExTensor, ExTensor, ExTensor, ExTensor, ExTensor,
    # Branch 3 param grads (same structure as branch 2 but 5x5)
    ExTensor, ExTensor, ExTensor, ExTensor, ExTensor, ExTensor, ExTensor, ExTensor,
    # Branch 4 param grads: proj conv w/b, bn gamma/beta
    ExTensor, ExTensor, ExTensor, ExTensor,
]:
```

Internally:
1. **Split** `grad_concat_out` into 4 slices at offsets `[0, c1, c1+c2, c1+c2+c3]` using `ExTensor.slice()` on axis=1
2. **Branch 1 backward**: `relu_backward → batch_norm2d_backward → conv2d_backward`
3. **Branch 2 backward**: `relu_backward → batch_norm2d_backward(bn3x3) → conv2d_backward(conv3x3) → relu_backward → batch_norm2d_backward(bn1x1_2) → conv2d_backward(conv1x1_2)`
4. **Branch 3 backward**: Same structure as branch 2 with `conv5x5` (padding=2)
5. **Branch 4 backward**: `relu_backward → batch_norm2d_backward(bn1x1_4) → conv2d_backward(conv1x1_4) → maxpool2d_backward`
6. **Combine** input gradients: `grad_x = grad_x_b1 + grad_x_b2 + grad_x_b3 + grad_x_b4` using element-wise add

#### 4e. `fn compute_gradients` function (replaces the placeholder closure)

The main function that:
1. Calls `forward_with_cache(model, batch_images)` → logits + cache
2. Calls `cross_entropy_backward` → grad_logits
3. Calls `linear_backward(grad_logits, cache.flattened_out, model.fc_weights)` → grad FC
4. Calls `dropout_backward` with saved mask
5. Calls `global_avgpool2d_backward` → grad after GAP
6. Unflatten gradient to spatial shape
7. Calls `inception_module_backward` for `inception_5b` → grad + param grads
8. Calls `inception_module_backward` for `inception_5a` → grad + param grads
9. Calls `maxpool2d_backward` for pool3
10. Calls `inception_module_backward` for `inception_4e` through `inception_4a` (5 modules)
11. Calls `maxpool2d_backward` for pool2
12. Calls `inception_module_backward` for `inception_3b` and `inception_3a`
13. Calls `maxpool2d_backward` for pool1
14. Initial conv block backward: `relu_backward → batch_norm2d_backward → conv2d_backward`
15. SGD+momentum update for all ~200 parameters using velocities list

#### 4f. Refactor `train_epoch` to use `compute_gradients`

Replace the `TrainingLoop` closure pattern with direct manual iteration (like AlexNet), passing `mut velocities: List[ExTensor]`:

```mojo
fn train_epoch(
    mut model: GoogLeNet,
    train_images: ExTensor,
    train_labels: ExTensor,
    batch_size: Int,
    learning_rate: Float32,
    momentum: Float32,
    epoch: Int,
    mut velocities: List[ExTensor],
) raises -> Float32:
```

#### 4g. Initialize velocities in `main()`

Before the training loop, initialize `List[ExTensor]` of ~200 zero tensors matching each parameter's shape. Order must match `compute_gradients`'s update order.

#### 4h. Remove placeholder print in `main()`

Delete lines 436-441 (the "NOTE: Full backward pass..." print block).

---

## 5. Implementation Order

1. **Read `shared/core/extensor.mojo`** — Verify `ExTensor.slice(start, end, axis)` signature and semantics for gradient splitting

2. **Read `shared/core/pooling.mojo`** — Verify `global_avgpool2d_backward` signature

3. **Define `InceptionModuleCache` struct** at top of `train.mojo` (before existing `train_epoch`)

4. **Define `GoogLeNetCache` struct** after `InceptionModuleCache`

5. **Implement `forward_with_cache`** — Run the full GoogLeNet forward pass, saving all intermediates into the cache structs. Verify shapes with comments matching `model.mojo`'s shape annotations.

6. **Implement `inception_module_backward`** — Start with Branch 1 (simplest), then branches 2-4. Test each branch's gradient shape independently.

7. **Implement `compute_gradients`** — Wire up the full backward pass using `forward_with_cache` + the inception backward helper + final layer backwards.

8. **Initialize and thread velocities** — Add velocity initialization in `main()` and update `train_epoch` signature.

9. **Remove placeholder comments and print statements** — Clean up the old documentation block and `NOTE:` prints.

10. **Add backward tests to `tests/models/test_googlenet_layers.mojo`** — Layerwise gradient tests for the inception backward helper.

11. **Compile check** — Run `pixi run mojo -I . examples/googlenet-cifar10/train.mojo` to verify no compile errors.

12. **Run existing tests** — `pixi run mojo -I . tests/models/test_googlenet_layers.mojo`

---

## 6. Verification

### Compile-time

```bash
cd /home/mvillmow/Odyssey2 && pixi run mojo -I . examples/googlenet-cifar10/train.mojo
```

### Unit tests — `tests/models/test_googlenet_layers.mojo`

Add the following test functions to the existing test file:

- `test_inception_backward_output_shapes()` — Verify `inception_module_backward` returns input grad with same shape as the module's input (e.g., `[2, 64, 8, 8]`).
- `test_inception_backward_branch1_gradient_nonzero()` — Branch 1 (1×1 conv) gradient is non-zero for a non-trivial input.
- `test_inception_backward_concat_split_inverse()` — That `concatenate_depthwise` + channel split round-trips correctly.
- `test_inception_backward_gradient_accumulation()` — Sum of branch input gradients has the correct shape.

### Integration check (small batch smoke test)

Add a minimal `test_train_step()` in `examples/googlenet-cifar10/test_model.mojo`:

```mojo
fn test_train_step() raises:
    """Verify one training step runs without error and loss is finite."""
    var model = GoogLeNet(num_classes=10)
    var input = zeros([4, 3, 32, 32], DType.float32)   # batch=4
    var labels = zeros([4], DType.uint8)
    var velocities = initialize_velocities(model)
    var loss = compute_gradients(model, input, labels, 0.01, 0.9, velocities)
    assert loss >= 0.0 and loss < 1000.0
```

### Full test suite

```bash
just test-group tests/models "test_googlenet*.mojo"
```

---

## 7. Critical Mojo Patterns (from Prior Learnings)

| Pattern | Rule |
|---|---|
| `batch_norm2d_backward` return | `Tuple[ExTensor, ExTensor, ExTensor]` — destructure as `var (gx, gg, gb) = batch_norm2d_backward(...)` |
| `conv2d_backward` return | `GradientTriple` — access `.grad_input`, `.grad_weights`, `.grad_bias` |
| ExTensor ownership | Use `^` when moving into function args or list writes; use `.copy()` when reading from list |
| List[ExTensor] aliasing | Extract with `var local = list[i]` before accessing `._data` or passing to ops |
| Tuple unpacking | Never `var a, b = fn()` with ExTensor — use `var r = fn(); var a = r[0]; var b = r[1]` |
| Closure captures | Cannot capture `mut model` (non-copyable) in closure — use direct function call pattern like AlexNet |
| SGD update | `vel = momentum * vel + lr * grad; weight -= vel` |
| Float type | Use `_get_float32()` not `_get_float64()` on Float32 tensors |

---

## 8. Velocity Tensor Count

Per `InceptionModule`: 4 branches × (conv_w, conv_b, bn_gamma, bn_beta) = ~20 params per module except branch 2 and 3 have 2 conv each + 2 BN = (2+2+2+2)×2+2+2 = 20 per module.

- 9 modules × ~20 params = 180 velocity tensors
- Initial conv: conv_w, conv_b, bn_gamma, bn_beta = 4
- FC: fc_w, fc_b = 2
- **Total: ~186 velocity tensors**

---

## 9. Skills Used

### Skills invoked during planning

- **`gh-read-issue-context`** (conceptually — reading issue #3295 context provided in prompt)

### Team knowledge base skills referenced (from Prior Learnings section)

| Skill | Category | How it influenced the plan |
|---|---|---|
| `batch-norm-backward-gradient-analysis` | testing | Confirmed `Tuple[ExTensor,ExTensor,ExTensor]` return type for `batch_norm2d_backward`; GradientTriple field names |
| `fix-implicitlycopyable-removal` | debugging | Drives all ownership patterns: `.copy()` from lists, `^` transfers, no closure captures of model |
| `fix-ci-memory-safety-failures` | debugging | Drives `var local = list[i]` pattern before `._data` access; avoids pointer-through-list-copy segfaults |
| `investigate-mojo-heap-corruption` | debugging | Keeps integration test count < 15; use `_get_float32()` not `_get_float64()` |
| `split-large-test-file` | testing | Plan avoids growing test file beyond ~3000 lines; backward tests added to existing file |
| `fix-placeholder-code-ci` | testing | Drives complete removal of ALL placeholder code (comments + code using placeholder results together) |
| `train-model` | training | SGD+momentum update formula: `vel = m*vel + lr*grad; w -= vel` |
| `fixme-todo-cleanup` | debugging | Remove `@fieldwise_init` if any cache struct uses custom `__init__` |

---
*Generated by Claude Code Planner*

**Comment by mvillmow** [comment]
Duplicate of #3181/#3184/#3187 — closing as part of bulk dedup
